### PR TITLE
[Azure Service Bus] Remove unnecessary LINQ on AmqpReceiver

### PR DESCRIPTION
**Pull request by danielmarbach**
Removes LINQ usage from the AmqpReceiver from the hot path to improve speed and allocations

I tried to document reasoning inline

## Results

``` ini

BenchmarkDotNet=v0.12.0, OS=Windows 10.0.18363
Intel Core i7-8550U CPU 1.80GHz (Kaby Lake R), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.1.201
  [Host]    : .NET Core 3.1.3 (CoreCLR 4.700.20.11803, CoreFX 4.700.20.12001), X64 RyuJIT
  MediumRun : .NET Core 3.1.3 (CoreCLR 4.700.20.11803, CoreFX 4.700.20.12001), X64 RyuJIT

Job=MediumRun  IterationCount=15  LaunchCount=2  
WarmupCount=10  

```
| Method | Elements |       Mean |     Error |    StdDev | Ratio | RatioSD |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------- |--------- |-----------:|----------:|----------:|------:|--------:|-------:|------:|------:|----------:|
| **Before** |        **1** |   **829.1 ns** |  **50.23 ns** |  **75.18 ns** |  **1.00** |    **0.00** | **0.0362** |     **-** |     **-** |     **152 B** |
|  After |        1 |   454.1 ns |  62.09 ns |  91.01 ns |  0.55 |    0.13 | 0.0286 |     - |     - |     120 B |
|        |          |            |           |           |       |         |        |       |       |           |
| **Before** |        **2** |   **693.1 ns** | **111.54 ns** | **166.95 ns** |  **1.00** |    **0.00** | **0.0401** |     **-** |     **-** |     **168 B** |
|  After |        2 |   567.6 ns |  90.46 ns | 129.74 ns |  0.87 |    0.29 | 0.0324 |     - |     - |     136 B |
|        |          |            |           |           |       |         |        |       |       |           |
| **Before** |        **4** |   **846.9 ns** |  **25.89 ns** |  **38.74 ns** |  **1.00** |    **0.00** | **0.0458** |     **-** |     **-** |     **200 B** |
|  After |        4 |   776.0 ns |  32.62 ns |  44.65 ns |  0.91 |    0.07 | 0.0401 |     - |     - |     168 B |
|        |          |            |           |           |       |         |        |       |       |           |
| **Before** |        **8** | **1,864.3 ns** | **228.77 ns** | **328.10 ns** |  **1.00** |    **0.00** | **0.0610** |     **-** |     **-** |     **264 B** |
|  After |        8 | 1,484.4 ns | 136.14 ns | 186.34 ns |  0.81 |    0.13 | 0.0553 |     - |     - |     232 B |
|        |          |            |           |           |       |         |        |       |       |           |
| **Before** |       **16** | **3,529.6 ns** | **459.86 ns** | **674.05 ns** |  **1.00** |    **0.00** | **0.0916** |     **-** |     **-** |     **392 B** |
|  After |       16 | 3,158.8 ns | 430.75 ns | 631.38 ns |  0.91 |    0.21 | 0.0839 |     - |     - |     360 B |

Benchmark used https://github.com/danielmarbach/MicroBenchmarks/blob/438dac249a7a44e178b758e860cf16ca834b0c79/MicroBenchmarks/Linq/LinqGuidConversionContains.cs

### CODE REVIEW

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs
Comment (danielmarbach) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs:297]: The idea is that we can avoid having to do the `Contains` if we already know we are in the request response scenario

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs
Comment (danielmarbach) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs:449]: Any not needed

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs
Comment (danielmarbach) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs:503]: Any not needed

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs
Comment (danielmarbach) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs:579]: Any not needed

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs
Comment (danielmarbach) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs:649]: We could also pass in the `List<Guid>` to avoid the cast if you prefer

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs
Comment (danielmarbach) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs:934]: I tried to do the necessary research in the AMQP lib used but couldn't figure out if the type here really has to be an array. I was hopping the tests would capture it

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs
Comment (danielmarbach) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs:942]: If nothing is returned then now an index out of bound would be raised instead of an InvalidOperation exception. Happy to roll back if you don't like this but technically this is the last LINQ usage that is not necessary under the assumption GetValue allows me to get the date times as an array (again I tried to do the necessary research in the AMQP lib but couldn't find anything)

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs
Comment (danielmarbach) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs:1134]: Same assumptions here. I see no reason to copy because the list cannot be reused before the operation is done

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs
Comment (pakrym) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs:577]: List<>(1) allocates 2 objects instead of 1 for array.

**Comment by pakrym** [comment]
Is there a benchmark/profile showing the improvement?

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs
Comment (ShivangiReja) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs:933]: AMQP encoding is different for both the types. Service Bus service expects ArrayEncoding here, if we pass list service would throw an error.

Comment (ShivangiReja) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs:933]: Thanks for the investigation but looks like we need to revert this.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs
Comment (ShivangiReja) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs:1134]: Same as above. ListEncoding and ArrayEncoding are differnt and Service Bus service expects ArrayEncoding here so we need to revert this.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpSender.cs
Comment (ShivangiReja) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpSender.cs:464]: We need to revert this as well.

**Comment by danielmarbach** [comment]
@pakrym happy to provide that if you think this is worth pursuing. I'll revert first the other changes @shivangireja pointed out

**Comment by pakrym** [comment]
> @pakrym happy to provide that if you think this is worth pursuing.

It's always good to have an idea of before/after and how the improvement compares to overall performance.

**Comment by danielmarbach** [comment]
@pakrym added the benchmark results with the changed code that hopefully also addresses the other comments. I ended up with a combination of LINQ and foreach to avoid the Any which leads to very good results.

I'm executing a medium run as I post this and will update the post as soon as I have the results

I'll leave it up to you if you want to take this or not. I have no problems closing this.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpSender.cs
Comment (danielmarbach) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpSender.cs]: @ShivangiReja thanks for taking the time to educate me! Highly appreciated. How does the team make sure that this is not accidentally changed in the future? Passing an array seems to be a critical part for this library to operate and should almost be considered some type of public contract that cannot be broken. So I would have hoped a test would fail if I invalidate such important assumptions. Or at least have some kind of an approval test that verifies the output map would have been great I think.

Comment (ShivangiReja) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpSender.cs]: We do have live tests for Service Bus SDK which cover almost all the scenarios and this [test](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/servicebus/Azure.Messaging.ServiceBus/tests/Sender/SenderLiveTests.cs#L170) would fail if you try to pass list instead of array.

If you would like to setup and run the live tests, you can follow this [guide](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/servicebus/Azure.Messaging.ServiceBus/CONTRIBUTING.md#development-environment-setup).

Comment (danielmarbach) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpSender.cs]: Forgive my ignorance but why was the PR green then (before my force push)? Are those tests not run on your CI/CD environment?

Comment (ShivangiReja) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpSender.cs]: Yeah, currently live tests don't run as a part of CI, only unit tests run because there isn't a recording test framework for AMQP as of now.

You can setup and run live tests locally by following this [guide](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/servicebus/Azure.Messaging.ServiceBus/CONTRIBUTING.md#development-environment-setup).

**Review by ShivangiReja — APPROVED** [review]
Changes look good to me! Thank you @danielmarbach!

@pakrym Could you please take a look?

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs
Comment (pakrym) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs:292]: Why not get rid of this piece of linq too?

**Comment by AlexanderSher** [comment]
The best gain in the test results above is for the array of 4 elements and it is 175 ns. How many times this method is called?

**Review by pakrym — APPROVED** [review]

**Comment by JoshLove-msft** [comment]
@danielmarbach thanks for these performance fixes! They are greatly appreciated 😄 Out of curiosity, have you profiled any end to end scenarios like receiving or sending to identify any bottlenecks? If so, we would be interested to learn about your findings.

**Comment by AlexanderSher** [comment]
I'm sorry for being stubborn, but I'm still interested in the scenario we're optimizing. I've done a simple benchmark based on the https://github.com/danielmarbach/MicroBenchmarks/commit/7a474da91836932c59a5cec74344e789453bd02e#diff-5d50ca96fc4df16892b002db1d70e8c3:

```cs
private static async Task<Guid[]> LinqWithAwait(IEnumerable<string> lockTokens)
{
    var lockTokenGuids = lockTokens.Select(token => new Guid(token)).ToArray();
    if (lockTokenGuids.Any(lockToken => _requestResponseLockedMessages.Contains(lockToken)))
        return await Task.FromResult(lockTokenGuids).ConfigureAwait(false);

    return await Task.FromResult(Array.Empty<Guid>()).ConfigureAwait(false);
}

private static async Task<Guid[]> ForeachWithAwait(IEnumerable<string> lockTokens)
{
    var requestResponse = false;
    var lockTokenGuids = lockTokens.Select(token => new Guid(token)).ToArray();
    foreach (var tokenGuid in lockTokenGuids)
    {
        if (!requestResponse && _requestResponseLockedMessages.Contains(tokenGuid)) requestResponse = true;
    }

    if (requestResponse)
    {
        return await Task.FromResult(lockTokenGuids).ConfigureAwait(false);
    }

    return await Task.FromResult(Array.Empty<Guid>()).ConfigureAwait(false);
}

private static Task<Guid[]> LinqWithoutAwait(IEnumerable<string> lockTokens)
{
    var lockTokenGuids = lockTokens.Select(token => new Guid(token)).ToArray();
    if (lockTokenGuids.Any(lockToken => _requestResponseLockedMessages.Contains(lockToken)))
        return Task.FromResult(lockTokenGuids);

    return Task.FromResult(Array.Empty<Guid>());
}
```

Here are the results:
|           Method |           Job |       Runtime | Elements |       Mean |    Error |    StdDev | Ratio | RatioSD |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------- |-------------- |-------------- |--------- |-----------:|---------:|----------:|------:|--------:|-------:|------:|------:|----------:|
|    LinqWithAwait |      .NET 4.8 |      .NET 4.8 |        4 | 2,767.8 ns | 55.45 ns | 118.16 ns |  1.00 |    0.00 | 0.0687 |     - |     - |     449 B |
| ForeachWithAwait |      .NET 4.8 |      .NET 4.8 |        4 | 2,708.8 ns | 54.26 ns | 115.64 ns |  0.98 |    0.06 | 0.0648 |     - |     - |     417 B |
| LinqWithoutAwait |      .NET 4.8 |      .NET 4.8 |        4 | 2,608.3 ns | 52.26 ns | 147.39 ns |  0.95 |    0.07 | 0.0572 |     - |     - |     369 B |
|                  |               |               |          |            |          |           |       |         |        |       |       |           |
|    LinqWithAwait | .NET Core 2.1 | .NET Core 2.1 |        4 | 1,933.2 ns | 38.37 ns |  71.11 ns |  1.00 |    0.00 | 0.0648 |     - |     - |     416 B |
| ForeachWithAwait | .NET Core 2.1 | .NET Core 2.1 |        4 | 1,870.9 ns | 37.34 ns | 107.74 ns |  0.97 |    0.05 | 0.0572 |     - |     - |     384 B |
| LinqWithoutAwait | .NET Core 2.1 | .NET Core 2.1 |        4 | 1,997.0 ns | 54.25 ns | 159.97 ns |  0.98 |    0.06 | 0.0534 |     - |     - |     344 B |
|                  |               |               |          |            |          |           |       |         |        |       |       |           |
|    LinqWithAwait | .NET Core 3.1 | .NET Core 3.1 |        4 | 1,061.0 ns | 23.19 ns |  68.36 ns |  1.00 |    0.00 | 0.0515 |     - |     - |     416 B |
| ForeachWithAwait | .NET Core 3.1 | .NET Core 3.1 |        4 |   986.7 ns | 19.76 ns |  57.63 ns |  0.94 |    0.08 | 0.0477 |     - |     - |     384 B |
| LinqWithoutAwait | .NET Core 3.1 | .NET Core 3.1 |        4 |   974.0 ns | 19.89 ns |  55.46 ns |  0.92 |    0.06 | 0.0439 |     - |     - |     344 B |

As you can see, simple removal of `async/await` in private method provides better results than LINQ replaced with `foreach`, both in speed and memory allocation. Yet we are still talking about nanoseconds and bytes here. This code can be improved even further, with total removal of LINQ and casting to array so that `foreach` doesn't allocate any memory, but will anyone see these optimizations in the shadow of `DisposeMessageRequestResponseAsync`?

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs
Comment (pakrym) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs:295]: Why don't you break loop early in that case?

Comment (danielmarbach) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs:295]: Because my brain was still held captive in the previous version of the code. Stupid me. Will fix

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs
Comment (danielmarbach) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs:292]: I did an attempt to do it by creating an array of guid. I then needed to ask for the count on the enumerable which lead to less allocations as the original version but worse performance

**Comment by danielmarbach** [comment]
@pakrym I added the break version, new results

``` ini

BenchmarkDotNet=v0.12.0, OS=Windows 10.0.18363
Intel Core i7-8550U CPU 1.80GHz (Kaby Lake R), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.1.201
  [Host]    : .NET Core 3.1.3 (CoreCLR 4.700.20.11803, CoreFX 4.700.20.12001), X64 RyuJIT
  MediumRun : .NET Core 3.1.3 (CoreCLR 4.700.20.11803, CoreFX 4.700.20.12001), X64 RyuJIT

Job=MediumRun  IterationCount=15  LaunchCount=2  
WarmupCount=10  

```
| Method | Elements |       Mean |     Error |    StdDev | Ratio | RatioSD |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------- |--------- |-----------:|----------:|----------:|------:|--------:|-------:|------:|------:|----------:|
| **Before** |        **1** |   **829.1 ns** |  **50.23 ns** |  **75.18 ns** |  **1.00** |    **0.00** | **0.0362** |     **-** |     **-** |     **152 B** |
|  After |        1 |   454.1 ns |  62.09 ns |  91.01 ns |  0.55 |    0.13 | 0.0286 |     - |     - |     120 B |
|        |          |            |           |           |       |         |        |       |       |           |
| **Before** |        **2** |   **693.1 ns** | **111.54 ns** | **166.95 ns** |  **1.00** |    **0.00** | **0.0401** |     **-** |     **-** |     **168 B** |
|  After |        2 |   567.6 ns |  90.46 ns | 129.74 ns |  0.87 |    0.29 | 0.0324 |     - |     - |     136 B |
|        |          |            |           |           |       |         |        |       |       |           |
| **Before** |        **4** |   **846.9 ns** |  **25.89 ns** |  **38.74 ns** |  **1.00** |    **0.00** | **0.0458** |     **-** |     **-** |     **200 B** |
|  After |        4 |   776.0 ns |  32.62 ns |  44.65 ns |  0.91 |    0.07 | 0.0401 |     - |     - |     168 B |
|        |          |            |           |           |       |         |        |       |       |           |
| **Before** |        **8** | **1,864.3 ns** | **228.77 ns** | **328.10 ns** |  **1.00** |    **0.00** | **0.0610** |     **-** |     **-** |     **264 B** |
|  After |        8 | 1,484.4 ns | 136.14 ns | 186.34 ns |  0.81 |    0.13 | 0.0553 |     - |     - |     232 B |
|        |          |            |           |           |       |         |        |       |       |           |
| **Before** |       **16** | **3,529.6 ns** | **459.86 ns** | **674.05 ns** |  **1.00** |    **0.00** | **0.0916** |     **-** |     **-** |     **392 B** |
|  After |       16 | 3,158.8 ns | 430.75 ns | 631.38 ns |  0.91 |    0.21 | 0.0839 |     - |     - |     360 B |

**Comment by danielmarbach** [comment]
@AlexanderSher It is true that in comparison to an IO-bound method this could be seen as unnecessary because one could argue the linq version could have better readability. What I have seen though for example with rabbitmq when fetching/sending messages under load those bytes can count. 

@JoshLove-msft I rebased to get your other fixes. For other PRs is there a way for me to have a full run done as part of the PR? It seems my previous PR was green but did cause problems when the pipeline run on master.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs
Comment (pakrym) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs:297]: You can inline the 

```
await DisposeMessageRequestResponseAsync();
return;

```

under the if check now, removing the extra variable.

Comment (danielmarbach) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs:297]: My preference was to keep the IO stuff out of the loop. I have no problems going your way if that is what you as maintainers prefer. Can only do in a few hours or I think you have edits rights on this PR

Comment (pakrym) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs:297]: I'll leave it to @ShivangiReja and @JoshLove-msft

**Comment by JoshLove-msft** [comment]
> @JoshLove-msft I rebased to get your other fixes. For other PRs is there a way for me to have a full run done as part of the PR? It seems my previous PR was green but did cause problems when the pipeline run on master.

Unfortunately, this would only be caught by the live tests which won't run automatically. I think it is on us at Microsoft to make sure we run the live tests before merging.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs:297]: I don't have a strong preference. I will note that we already are following the pattern of returning from at multiple points in this method, which I find a bit harder to read, especially since we unconditionally call `DisposeMessagesAsync` at the end of the method. I would prefer if we had an else just to make it more clear.

Comment (ShivangiReja) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs:297]: I like the way @pakrym suggested. If we would move `DisposeMessageRequestResponseAsync` inside the if block we can get rid of the extra variable.

Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs:297]: I can live with that.

Comment (danielmarbach) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs]: Fixed

**Comment by JoshLove-msft** [comment]
It looks like all the comments have been settled. I'm planning on merging this in assuming the live tests pass.

**Comment by JoshLove-msft** [comment]
@danielmarbach looks like there is a conflict now with something that I merged in. Sorry about that. Would you mind resolving this?

**Comment by danielmarbach** [comment]
Rebased

**Pull request merged**
